core/logtailer/tailer.go

```diff
@@ -0,0 +1,400 @@
+			if err != nil {
+				if deserialisationFailures == 0 {
+					logger.Warningf("log deserialization failed, %v", err)
+				}
+				deserialisationFailures++
+				continue
+			} else {
+				if deserialisationFailures > 1 {
+					logger.Debugf("total of %d log serialisation errors", deserialisationFailures)
+				}
+				deserialisationFailures = 0
+			}
+			if !t.includeRecord(rec) {
+				continue
+			}
+			select {
+			case <-t.tomb.Dying():
+				return tomb.ErrDying
+			case t.logCh <- rec:
+			}
+		}
+	}
+}
+
+func (t *logTailer) includeRecord(rec *corelogger.LogRecord) bool {
+	if rec.Time.Before(t.params.StartTime) {
+		return false
+	}
+	if rec.Level < t.params.MinLevel {
+		return false
+	}
+	if len(t.params.IncludeEntity) > 0 {
+		match, err := regexp.MatchString(makeEntityPattern(t.params.IncludeEntity), rec.Entity)
+		if !match || err != nil {
+			return false
+		}
+	}
+	if len(t.params.ExcludeEntity) > 0 {
+		match, err := regexp.MatchString(makeEntityPattern(t.params.ExcludeEntity), rec.Entity)
+		if match || err != nil {
+			return false
+		}
+	}
+	if len(t.params.IncludeModule) > 0 {
+		match, err := regexp.MatchString(makeModulePattern(t.params.IncludeModule), rec.Module)
+		if !match || err != nil {
+			return false
+		}
+	}
+	if len(t.params.ExcludeModule) > 0 {
+		match, err := regexp.MatchString(makeModulePattern(t.params.ExcludeModule), rec.Module)
+		if match || err != nil {
+			return false
+		}
+	}
+	// TODO(debug-log) - implement label filtering when json formatting is added
+	if len(t.params.IncludeLabel) > 0 {
+	}
+	if len(t.params.ExcludeLabel) > 0 {
+	}
+	return true
+}
+
+func makeEntityPattern(entities []string) string {
+	var patterns []string
+	for _, entity := range entities {
+		// Convert * wildcard to the regex equivalent. This is safe
+		// because * never appears in entity names. Escape any other regex.
+		escaped := regexp.QuoteMeta(entity)
+		unescapedWildcards := strings.Replace(escaped, regexp.QuoteMeta("*"), ".*", -1)
+		patterns = append(patterns, unescapedWildcards)
+	}
+	return `^(` + strings.Join(patterns, "|") + `)$`
+}
+
+func makeModulePattern(modules []string) string {
+	var patterns []string
+	for _, module := range modules {
+		patterns = append(patterns, regexp.QuoteMeta(module))
+	}
+	return `^(` + strings.Join(patterns, "|") + `)(\..+)?$`
+}
+
+func logLineToRecord(modelUUID string, line string) (*corelogger.LogRecord, error) {
+	parts := strings.SplitN(line, " ", 7)
+	if len(parts) < 7 {
+		return nil, errors.Errorf("invalid log line %q", line)
+	}
+	level, ok := loggo.ParseLevel(parts[3])
+	if !ok {
+		return nil, errors.Errorf("unrecognized log level %q", parts[3])
+	}
+	timeStr := parts[1] + " " + parts[2]
+	timeStamp, err := time.Parse("2006-01-02 15:04:05", timeStr)
+	if err != nil {
+		return nil, errors.Annotatef(err, "invalid log timestamp %q", timeStr)
+	}
+
+	rec := &corelogger.LogRecord{
+		Time: timeStamp.UTC(), // not worth preserving TZ
```

Comment:
Everything should be UTC anyway.

Reply:
Yes. This was a carry over from existing code which I'll tweak.